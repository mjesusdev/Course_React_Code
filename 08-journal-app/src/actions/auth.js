import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';

// Sign in with Email & Password
export const startLoginEmailPassword = ( email, password ) => {
    return ( dispatch ) => {
        setTimeout(() => {
            dispatch( login(123, 'Peter') );
        }, 3000)
    }
}

// Register User with Email & Password
export const startRegisterUser = ( email, password, name ) => {
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password)
            .then( async ({ user }) => {

                await user.updateProfile({ displayName: name });

                dispatch(
                    login( user.uid, user.displayName )
                )
            })

            .catch( error => {
                console.log(error);
            })
    }
}

// Sign in with Google Account
export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });
    }
}

// Export
export const login = ( uid, displayName ) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})