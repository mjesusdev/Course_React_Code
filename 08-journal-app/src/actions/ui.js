import { types } from '../types/types';

export const setError = ( err ) => ({
    type: types.uiSetError,
    payload: err
});

export const removeError = ( err ) => ({
    type: types.uiRemoveError,
    payload: err
});

export const startLoading = () => ({
    type: types.uiStartLoading
})

export const finishLoading = () => ({
    type: types.uiFinishLoading
})