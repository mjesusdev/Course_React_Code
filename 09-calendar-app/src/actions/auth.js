import { fetchWithToken, fetchWitoutToken } from "../helpers/fetch"
import { types } from "../types/types";
import Swal from 'sweetalert2';

export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        const resp = await fetchWitoutToken( 'auth', { email, password }, 'POST' );
        const body = await resp.json();
        
        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startRegister = ( email, password, name ) => {
    return async( dispatch ) => {
        const resp = await fetchWitoutToken('auth/new', { email, password, name }, 'POST' );
        const body = await resp.json();

        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( register({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const startChecking = () => {
    return async ( dispatch ) => {
        const resp = await fetchWithToken( 'auth/renew' );
        const body = await resp.json();
        
        if ( body.ok ) {
            // Save Token in LocalStorage ✅
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );
        
            dispatch( register({
                uid: body.uid,
                name: body.name
            }) );
        } else {
            dispatch( checkingFinish() );
        }
    }
}

const checkingFinish = () => ({
    type: types.authCheckingFinish
})

const login = ( user ) => ({ 
    type: types.authLogin,
    payload: user
});

const register = ( user ) => ({
    type: types.authStartRegister,
    payload: user
})