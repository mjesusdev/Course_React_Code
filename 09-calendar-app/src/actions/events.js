import { types } from "../types/types";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import Swal from "sweetalert2";

export const eventStartAddNew = ( event ) => {
    return async ( dispatch, getState ) => {
        
        const { uid, name } = getState().auth;

        try {
            const resp = await fetchWithToken('events', event, 'POST');
            const body = await resp.json();

            if ( body.ok ) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name: name
                }

                dispatch( eventAddNew(event) );
            }

        } catch (error) {
            console.log(error);
        }

    }
}

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = ( event ) => {
    return async ( dispatch ) => {
        try {
            const resp = await fetchWithToken(`events/${ event.id }`, event, 'PUT');
            const body = await resp.json();

            if ( body.ok ) {
                dispatch( eventUpdated( event ) );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);    
        }
    }
}

const eventUpdated = ( event ) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = ( event ) => {
    return async ( dispatch, getState ) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const resp = await fetchWithToken(`events/${ id }`, event, 'DELETE');
            const body = await resp.json();

            if ( body.ok ) {
                dispatch( eventDeleted() );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);    
        }
    }
}

const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventStartLoading = () => {
    return async ( dispatch ) => {
        try {
            
            const resp = await fetchWithToken('events');
            const body = await resp.json();
            
            const events = prepareEvents(body.events);
            dispatch( eventLoaded(events) );
            
        } catch (error) {
            console.log(error);
        }
    }
};

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({ 
    type: types.eventLogout 
});