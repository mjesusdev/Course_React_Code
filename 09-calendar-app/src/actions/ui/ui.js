import { types } from '../../types/types';

export const uiOpenModal = () => ({ 
    type: types.uiOpenModal 
});

export const uiCloseModal = () => ({ 
    type: types.uiCloseModal 
});