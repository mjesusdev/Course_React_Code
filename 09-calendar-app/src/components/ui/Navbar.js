import React from 'react'

export const Navbar = () => {
    return (
        <div className="navbar navbar-dark bg-dark mb-4">
            <span className="navbar-brand">
                Peter
            </span>

            <button className="btn btn-outline-danger">
                <i className="fas fa-sign-out-alt"></i>
                <span> Exit</span>
            </button>
        </div>
    )
}
