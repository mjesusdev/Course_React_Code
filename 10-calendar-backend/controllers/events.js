// Functions for events ✅
const { response } = require('express');
const Event = require('../models/Event');

const getEvents = async ( req, res = response ) => {

    const events = await Event.find()
                              .populate('user', 'name')

    res.json({
        ok: true,
        events
    });
}

const createEvent = async ( req, res = response ) => {

    const event = new Event( req.body );
    
    try {

        event.user = req.uid;

        const eventSave = await event.save();

        res.json({
            ok: true,
            event: eventSave
        })

    } catch(error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Speak with the Administrator, please!'
        })
    }
}

const updateEvent = async ( req, res = response ) => {

    const eventId = req.params.id;
    const uid = req.uid;

    try {

        const event = await Event.findById( eventId );

        if ( !event ) {
            return res.status(404).json({
                ok: false,
                msg: 'Event not exists with that ID ❗'
            });
        }

        if ( event.user.toString() !== uid ) {
            return res.status(401).json({
                ok: false,
                msg: `You don't have permission for edit this event ❗`
            })
        }

        const newEvent = {
            ...req.body,
            user: uid
        }

        const eventUpdated = await Event.findByIdAndUpdate( eventId, newEvent, { new: true } );

        res.json({
            ok: true,
            event: eventUpdated
        })

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Speak with the Administrator, please!'
        })
    }
}

const deleteEvent = async ( req, res = response ) => {

    const eventId = req.params.id;
    const uid = req.uid;

    try {
        const event = await Event.findById( eventId );

        if ( !event ) {
            return res.status(404).json({
                ok: false,
                msg: 'Event not exists with that ID ❗'
            });
        }

        if ( event.user.toString() !== uid ) {
            return res.status(401).json({
                ok: false,
                msg: 'You do not have permise for edit this event ❗'
            })
        }

        await Event.findByIdAndDelete( eventId );

        res.json({
            ok: true,
            msg: 'Event Deleted'
        })

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Speak with the Administrator, please!'
        })
    }
}

module.exports = {
    getEvents,
    createEvent,
    updateEvent,
    deleteEvent
}