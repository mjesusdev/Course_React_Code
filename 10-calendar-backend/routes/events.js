/* 
    Event Routes ✅
    /api/events
*/

const { Router } = require('express');
const { validateJWT } = require('../middlewares/validate-jwt');
const { getEvents, createEvent, updateEvent, deleteEvent } = require('../controllers/events');

const router = Router();

// Validate JWT - All routes 👏
router.use( validateJWT );

// Get Events
router.get('/', getEvents);

// Create a new Event
router.post('/', createEvent);

// Update Event
router.put('/:id', updateEvent);

// Delete Event
router.delete('/:id', deleteEvent);

module.exports = router;