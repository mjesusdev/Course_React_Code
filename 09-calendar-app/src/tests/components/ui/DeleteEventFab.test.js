import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';
import { eventStartDelete } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
    eventStartDelete: jest.fn()
}));

// Configure Thunk and Store
const middlewars = [thunk];
const mockStore = configureStore(middlewars);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <DeleteEventFab />
    </Provider>
)

describe('Test in <DeleteEventFab />', () => {

    test('should shows correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should run eventStartDelete when do click', () => {
        wrapper.find('button').simulate('click');

        expect( eventStartDelete ).toHaveBeenCalled();
    })
    
    
})
