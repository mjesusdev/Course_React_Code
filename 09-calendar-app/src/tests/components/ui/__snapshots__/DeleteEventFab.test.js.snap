// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <DeleteEventFab /> should shows correctly 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Delete Event 
  </span>
</button>
`;
